addons/sale_project/models/account_move_line.py

```diff
@@ -13,6 +14,45 @@ def _compute_analytic_distribution(self):
         super(AccountMoveLine, self - project_amls)._compute_analytic_distribution()
         project_id = self._context.get('project_id', False)
         if project_id:
-            analytic_account = self.env['project.project'].browse(project_id).analytic_account_id
+            project = self.env['project.project'].browse(project_id)
             for line in self:
-                line.analytic_distribution = {analytic_account.id: 100}
+                line.analytic_distribution = project._get_analytic_distribution()
+
+    def _get_so_mapping_from_project(self):
+        """ Get the mapping of move.line with the sale.order record on which its analytic entries should be reinvoiced.
+            A sale.order matches a move.line if the sale.order's project contains all the same analytic accounts
+            as the ones in the distribution of the move.line.
+            :return a dict where key is the move line id, and value is sale.order record (or None).
+        """
+        mapping = {}
+        ProjectProject = self.env['project.project']
+        AnalyticAccount = self.env['account.analytic.account']
+        for move_line in self:
+            if not move_line.analytic_distribution:
+                continue
+            distribution_json = move_line.analytic_distribution
+            project_query = ProjectProject._search(OR([
+                AND([
+                    [(AnalyticAccount.browse(int(account_id)).root_plan_id._column_name(), "=", int(account_id))]
+                    for account_id in key.split(",")
+                ])
+                for key in distribution_json
+            ]))
+
+            sale_order = self.env['sale.order'].search(
+                [("project_id", 'in', project_query), ('state', '=', 'sale')],
+                order='create_date ASC', limit=1,
+            )
+            if sale_order:
+                mapping[move_line.id] = sale_order
+            else:
+                sale_order = self.env['sale.order'].search([("project_id", 'in', project_query)], order='create_date ASC', limit=1)
+                mapping[move_line.id] = sale_order
+
+        # map of AAL index with the SO on which it needs to be reinvoiced. Maybe be None if no SO found
+        return mapping
```

Comment:
is it not possible to do the query in batch instead of per move line?

Reply:
Done :) Can you check again?